packages/flutter/lib/src/rendering/sliver.dart

```diff
@@ -1742,46 +1440,599 @@ class RenderViewport2 extends RenderBox with ContainerRenderObjectMixin<RenderSl
+      crossExtent,
+      GrowthDirection.forward,
+      childAfter,
+    );
+  }
+
+  @override
+  bool get hasVisualOverflow => _hasVisualOverflow;
+
+  @override
+  void updateOutOfBoundsData(GrowthDirection growthDirection, SliverGeometry childLayoutGeometry) {
+    switch (growthDirection) {
+      case GrowthDirection.forward:
+        _maxScrollExtent += childLayoutGeometry.scrollExtent;
+        break;
+      case GrowthDirection.reverse:
+        _minScrollExtent -= childLayoutGeometry.scrollExtent;
+        break;
+    }
+    if (childLayoutGeometry.hasVisualOverflow)
+      _hasVisualOverflow = true;
+  }
+
+  @override
+  void updateChildPaintOffset(RenderSliver child, double paintOffset, GrowthDirection growthDirection) {
+    final SliverPhysicalParentData childParentData = child.parentData;
+    childParentData.paintOffset = computeAbsolutePaintOffset(child, paintOffset, growthDirection);
+  }
 
+  @override
+  Offset paintOffsetOf(RenderSliver child) {
+    final SliverPhysicalParentData childParentData = child.parentData;
+    return childParentData.paintOffset;
+  }
+
+  @override
+  void applyPaintTransform(RenderObject child, Matrix4 transform) {
+    assert(child != null);
+    final SliverPhysicalParentData childParentData = child.parentData;
+    childParentData.applyPaintTransform(transform);
+  }
+
+  @override
+  double computeChildMainAxisPosition(RenderSliver child, double parentMainAxisPosition) {
+    assert(child != null);
+    assert(child.constraints != null);
+    final SliverPhysicalParentData childParentData = child.parentData;
+    switch (applyGrowthDirectionToAxisDirection(child.constraints.axisDirection, child.constraints.growthDirection)) {
+      case AxisDirection.up:
+        return child.geometry.paintExtent - (parentMainAxisPosition - childParentData.paintOffset.dy);
+      case AxisDirection.right:
+        return parentMainAxisPosition - childParentData.paintOffset.dx;
+      case AxisDirection.down:
+        return parentMainAxisPosition - childParentData.paintOffset.dy;
+      case AxisDirection.left:
+        return child.geometry.paintExtent - (parentMainAxisPosition - childParentData.paintOffset.dx);
+    }
+    return 0.0;
+  }
+
+  @override
+  void rereportDimensions() {
+    assert(_minScrollExtent != null);
+    assert(_maxScrollExtent != null);
+    assert(anchor != null);
+    double effectiveExtent;
+    switch (axis) {
+      case Axis.vertical:
+        effectiveExtent = size.height;
+        break;
+      case Axis.horizontal:
+        effectiveExtent = size.width;
+        break;
+    }
+    assert(effectiveExtent != null);
+    offset.applyViewportDimension(effectiveExtent);
+    if (!offset.applyContentDimensions(
+          // when updating this, also update similar code in performLayout()
+          math.min(0.0, _minScrollExtent + effectiveExtent * anchor),
+          math.max(0.0, _maxScrollExtent - effectiveExtent * (1.0 - anchor)),
+       ))
+     markNeedsLayout();
```

Comment:
I wasn't expecting this function to call markNeedsLayout based on the name...  Maybe have this function return a bool and have the caller call markNeedsLayout?